### PR TITLE
rootless: fix exec

### PR DESCRIPTION
**Pull request by giuseppe**
We cannot re-exec into a new user namespace to gain privileges and
access an existing as the new namespace is not the owner of the
existing container.

"unshare" is used to join the user namespace of the target container.

The current implementation assumes that the main process of the
container didn't create a new user namespace.

Since in the setup phase we are not running with euid=0, we must skip
the setup for containers/storage.

Closes: https://github.com/containers/libpod/issues/1329
Closes: https://github.com/containers/libpod/issues/1337

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

File: libpod/runtime.go [hunk 1]
```diff
@@ -406,7 +409,7 @@ func makeRuntime(runtime *Runtime) (err error) {
 	// Setting signaturepolicypath
 	ir.SignaturePolicyPath = runtime.config.SignaturePolicyPath
 	defer func() {
-		if err != nil {
+		if err != nil && store != nil {
```
Comment (mheon) [inline libpod/runtime.go]: Will the nil check work here? Store is an interface, not a pointer, so you usually can't check against nil directly

**Comment by giuseppe** [comment]
@mheon I've looked more but I don't see any other nice way of solving it :(

**Comment by mheon** [comment]
@giuseppe Alright, fair enough. Let's review and merge for release, then

File: libpod/runtime.go [hunk 2]
```diff
@@ -398,9 +398,12 @@ func makeRuntime(runtime *Runtime) (err error) {
 	}
 
 	// Set up containers/storage
-	store, err := storage.GetStore(runtime.config.StorageConfig)
-	if err != nil {
-		return err
+	var store storage.Store
+	if !rootless.SkipStorageSetup() {
```
[the remark this answers is not in the document]
Comment (giuseppe) [reply libpod/runtime.go]: Done.

I think this change makes sense in general, as we don't need to try to setup the storage when `exec'uting` a command but I didn't feel confident to make the change as part of this PR :-)

**Comment by mheon** [comment]
@rhatdan @vrothberg @baude PTAL

**Comment by mheon** [comment]
@giuseppe Cross-build errors in Travis

**Comment by rhatdan** [comment]
LGTM

File: libpod/runtime.go [hunk 3]
```diff
@@ -398,9 +398,13 @@ func makeRuntime(runtime *Runtime) (err error) {
...
+		logrus.Debug("Not configuring container store")
```
Comment (mheon) [inline libpod/runtime.go]: Can you move this into an else so it only prints if we're not configuring the store?

File: pkg/rootless/rootless_linux.go [hunk 4]
```diff
@@ -172,3 +188,86 @@ func BecomeRootInUserNS() (bool, int, error) {
 
 	return true, int(ret), nil
 }
+
+func readUserNs(path string) (string, error) {
+	b := make([]byte, 256)
```
Comment (mheon) [inline pkg/rootless/rootless_linux.go]: Are we guaranteed this will be big enough?

Comment (giuseppe) [reply pkg/rootless/rootless_linux.go]: that is much bigger than needed, the link is something like "user:[4026531837]".  Should I make the buffer smaller?

Comment (mheon) [reply pkg/rootless/rootless_linux.go]: Oh, I thought we were potentially holding a full path. 256 is probably still fine, we aren't exactly optimizing for memory usage yet.

**Comment by mheon** [comment]
Rootless tests seem to be failing on Travis

**Comment by giuseppe** [comment]
hm.. they pass locally, can't understand what is going wrong on Travis

**Comment by giuseppe** [comment]
Travis uses some ancient kernel that lacks the ioctl that is needed to lookup the parent of an user namespace.  I've added a check for that and now tests are happy.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**